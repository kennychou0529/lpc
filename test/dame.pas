program dame;
label 1;

type vec=array[1..8] of 0..255;
var x:vec;
  i,k:integer;

function f(x:vec;i:integer):boolean;
label 10;
var j:integer;
begin
f:=false;
for j:=1 to i-1 do if x[i]=x[j] then goto 10
                                 else if i-j=abs(x[i]-x[j]) then goto 10;
f:=true;
10:end;

begin
 for i:=1 to 8 do x[i]:=0;
 i:=1;
 while i>0 do begin
             k:=0;
             while x[i]<8 do begin
                           x[i]:=x[i]+1;
                           if f(x,i) then begin k:=1;goto 1;end;
                           end;
             1:if k=1 then if i=8 then begin
                                       writeln(' solutie : ');
                                       for i:=1 to 8 do write(x[i],'  ');

                                       writeln
                                       end
                                 else begin i:=i+1;x[i]:=0;end
                     else i:=i-1;
              end;
 end.