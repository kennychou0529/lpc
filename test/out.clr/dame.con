
compiling pass 1 (LPC)
-----------------------------------------------------
  warning on line 8: 'x' hiding existing declaration on line 5
  warning on line 8: 'i' hiding existing declaration on line 6

0 error(s), 2 warning(s)

Exit code: 0

compiling pass 2 (target = clr)
-----------------------------------------------------

Invoking ILASM ...
Exit code: 0

Verifying the compiled binary (PEVERIFY) ...
All Classes and Methods in dame.exe Verified.
Exit code: 0

running the executable
-----------------------------------------------------

Exit code: 0