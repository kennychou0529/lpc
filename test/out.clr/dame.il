
#include "commonDefs.il"

.field public static string[] _ProgramArgs

.method static void .cctor()
{
    ldc.i4 2
    newarr string
    dup
    ldc.i4 0
    ldstr "_input"
    stelem string
    dup
    ldc.i4 1
    ldstr "_output"
    stelem string
    stsfld string[] _ProgramArgs
    ret
}

// "dame.pas"

// line 8
//================================================================================
// scope: f (level : 2)

// line 0xfeefee

// TYPE vec = array [1 .. 8] of uint8;
.class value explicit T_vec
{
    // elem size = 1
    // alignment = 1
    .size 8

    .field [0] public uint8 elem

    .method public uint8& ldadr(int32 index)
    {
        // check(index >= 1)
        ldarg index
        ldc.i4 1
        blt badIndex
    
        // check(index <= 8)
        ldarg index
        ldc.i4 8
        bgt badIndex
    
        ldarg.0
        ldarg index
        ldc.i4 1
        sub
        add
        ret
    
    badIndex:
        ldstr "index out of bounds"
        newobj instance void [mscorlib]System.Exception::.ctor(string)
        throw
    }
}

.class @f
{
    // parameters
    .field public int32 i
    .field public valuetype T_vec x

    // local variables
    .field public bool _fnvalue
    .field public int32 j

    .method public void .ctor()
    {
        ldarg.0
        call instance void [mscorlib]System.Object::.ctor()
        ret
    }
}


// line 1
//================================================================================
// scope: program scope (level : 1)

// global variables
.class _G
{
    .field public static valuetype [lpcRuntime]LPC.File _input
    .field public static valuetype [lpcRuntime]LPC.File _output
    .field public static int32 i
    .field public static int32 k
    .field public static valuetype T_vec x
}

// line 8

.method static
bool f(valuetype T_vec x, int32 i)
{
    .maxstack 32

    .locals init(
        class @f _F,
        int32 $tmpEndValue1,
        int32 $tmpIndex2,
        int32 $tmpIndex3)
    
    newobj instance void @f::.ctor()
    stloc _F
    
    ldloc _F
    ldarg i
    stfld int32 @f::i
    
    ldloc _F
    ldarg x
    stfld valuetype T_vec @f::x

L_Entry:
    // line 12
    ldloc _F
    ldc.i4 0 // (false)
    stfld bool @f::_fnvalue
    // line 13
    ldloc _F
    ldfld int32 @f::i
    ldc.i4 1
    sub
    stloc $tmpEndValue1
    ldc.i4 1
    dup
    ldloc $tmpEndValue1
    bgt L_for_end_2
    stloc $tmpIndex2
    ldloc _F
    ldloc $tmpIndex2
    stfld int32 @f::j
L_for_loop_1:
    // line 13
    ldloc _F
    ldflda valuetype T_vec @f::x
    ldloc _F
    ldfld int32 @f::i
    call instance uint8& T_vec::ldadr(int32)
    ldobj uint8
    ldloc _F
    ldflda valuetype T_vec @f::x
    ldloc _F
    ldfld int32 @f::j
    call instance uint8& T_vec::ldadr(int32)
    ldobj uint8
    ceq
    brfalse L_else_3
    // line 13
    br L_10
    br L_if_end_4
L_else_3:
    // line 14
    ldloc _F
    ldfld int32 @f::i
    ldloc _F
    ldfld int32 @f::j
    sub
    ldloc _F
    ldflda valuetype T_vec @f::x
    ldloc _F
    ldfld int32 @f::i
    call instance uint8& T_vec::ldadr(int32)
    ldobj uint8
    ldloc _F
    ldflda valuetype T_vec @f::x
    ldloc _F
    ldfld int32 @f::j
    call instance uint8& T_vec::ldadr(int32)
    ldobj uint8
    sub
    dup
    ldc.i4 0
    bge L_abs_question_5
    neg
L_abs_question_5:
    ceq
    brfalse L_if_end_6
    // line 14
    br L_10
L_if_end_6:
L_if_end_4:
    ldloc _F
    ldfld int32 @f::j
    dup
    ldloc $tmpEndValue1
    beq L_for_end_2
    ldc.i4.1
    add
    stloc $tmpIndex3
    ldloc _F
    ldloc $tmpIndex3
    stfld int32 @f::j
    br L_for_loop_1
L_for_end_2:
    pop
    // line 15
    ldloc _F
    ldc.i4 1 // (true)
    stfld bool @f::_fnvalue
L_10:
    /* nop */

L_Epilogue:
    ldloc _F
    ldfld bool @f::_fnvalue
    ret
}

// line 0
//================================================================================
// scope: global scope (level : 0)

// line 1

.method static
void _Program()
{
    .maxstack 32

    .locals init(
        int32 $tmpEndValue1,
        int32 $tmpIndex2,
        int32 $tmpIndex3,
        int32 $tmpEndValue4,
        int32 $tmpIndex5,
        int32 $tmpIndex6)

    ldsflda valuetype [lpcRuntime]LPC.File _G::_input
    ldc.i4 0
    call instance void [lpcRuntime]LPC.File::Init(int32)

    ldsflda valuetype [lpcRuntime]LPC.File _G::_output
    ldc.i4 1
    call instance void [lpcRuntime]LPC.File::Init(int32)

L_Entry:
    // line 19
    ldc.i4 8
    stloc $tmpEndValue1
    ldc.i4 1
    dup
    ldloc $tmpEndValue1
    bgt L_for_end_2
    stloc $tmpIndex2
    ldloc $tmpIndex2
    stsfld int32 _G::i
L_for_loop_1:
    // line 19
    ldsflda valuetype T_vec _G::x
    ldsfld int32 _G::i
    call instance uint8& T_vec::ldadr(int32)
    ldc.i4 0
    dup
    ldc.i4 0
    ldc.i4 255
    call void [lpcRuntime]LPC.Runtime::RTCheckRange(int32, int32, int32)
    stobj uint8
    ldsfld int32 _G::i
    dup
    ldloc $tmpEndValue1
    beq L_for_end_2
    ldc.i4.1
    add
    stloc $tmpIndex3
    ldloc $tmpIndex3
    stsfld int32 _G::i
    br L_for_loop_1
L_for_end_2:
    pop
    // line 20
    ldc.i4 1
    stsfld int32 _G::i
    // line 21
    br L_while_test_4
L_while_3:
    // line 22
    ldc.i4 0
    stsfld int32 _G::k
    // line 23
    br L_while_test_6
L_while_5:
    // line 24
    ldsflda valuetype T_vec _G::x
    ldsfld int32 _G::i
    call instance uint8& T_vec::ldadr(int32)
    ldsflda valuetype T_vec _G::x
    ldsfld int32 _G::i
    call instance uint8& T_vec::ldadr(int32)
    ldobj uint8
    ldc.i4 1
    add
    dup
    ldc.i4 0
    ldc.i4 255
    call void [lpcRuntime]LPC.Runtime::RTCheckRange(int32, int32, int32)
    stobj uint8
    // line 25
    ldsfld valuetype T_vec _G::x
    ldsfld int32 _G::i
    call bool f(valuetype T_vec x, int32 i)
    brfalse L_if_end_7
    // line 25
    ldc.i4 1
    stsfld int32 _G::k
    // line 25
    br L_1
    /* nop */
L_if_end_7:
    /* nop */
L_while_test_6:
    ldsflda valuetype T_vec _G::x
    ldsfld int32 _G::i
    call instance uint8& T_vec::ldadr(int32)
    ldobj uint8
    ldc.i4 8
    clt
    brtrue L_while_5
    // line 27
L_1:
    ldsfld int32 _G::k
    ldc.i4 1
    ceq
    brfalse L_else_8
    // line 27
    ldsfld int32 _G::i
    ldc.i4 8
    ceq
    brfalse L_else_10
    // line 28
    ldsflda valuetype [lpcRuntime]LPC.File _G::_output
    dup
    ldstr " solutie : "
    call instance void [lpcRuntime]LPC.File::Write(string)
    call instance void [lpcRuntime]LPC.File::WriteLn()
    // line 29
    ldc.i4 8
    stloc $tmpEndValue4
    ldc.i4 1
    dup
    ldloc $tmpEndValue4
    bgt L_for_end_13
    stloc $tmpIndex5
    ldloc $tmpIndex5
    stsfld int32 _G::i
L_for_loop_12:
    // line 29
    ldsflda valuetype [lpcRuntime]LPC.File _G::_output
    dup
    ldsflda valuetype T_vec _G::x
    ldsfld int32 _G::i
    call instance uint8& T_vec::ldadr(int32)
    ldobj uint8
    call instance void [lpcRuntime]LPC.File::Write(int32)
    dup
    ldstr "  "
    call instance void [lpcRuntime]LPC.File::Write(string)
    pop
    ldsfld int32 _G::i
    dup
    ldloc $tmpEndValue4
    beq L_for_end_13
    ldc.i4.1
    add
    stloc $tmpIndex6
    ldloc $tmpIndex6
    stsfld int32 _G::i
    br L_for_loop_12
L_for_end_13:
    pop
    // line 31
    ldsflda valuetype [lpcRuntime]LPC.File _G::_output
    call instance void [lpcRuntime]LPC.File::WriteLn()
    br L_if_end_11
L_else_10:
    // line 33
    ldsfld int32 _G::i
    ldc.i4 1
    add
    stsfld int32 _G::i
    // line 33
    ldsflda valuetype T_vec _G::x
    ldsfld int32 _G::i
    call instance uint8& T_vec::ldadr(int32)
    ldc.i4 0
    dup
    ldc.i4 0
    ldc.i4 255
    call void [lpcRuntime]LPC.Runtime::RTCheckRange(int32, int32, int32)
    stobj uint8
    /* nop */
L_if_end_11:
    br L_if_end_9
L_else_8:
    // line 34
    ldsfld int32 _G::i
    ldc.i4 1
    sub
    stsfld int32 _G::i
L_if_end_9:
    /* nop */
L_while_test_4:
    ldsfld int32 _G::i
    ldc.i4 0
    cgt
    brtrue L_while_3
    /* nop */
    leave L_Epilogue

L_Epilogue:
    ret

.try L_Entry to L_Epilogue finally
{
    ldsflda valuetype [lpcRuntime]LPC.File _G::_input
    call instance void [lpcRuntime]LPC.File::Close()

    ldsflda valuetype [lpcRuntime]LPC.File _G::_output
    call instance void [lpcRuntime]LPC.File::Close()

    endfinally
}

}
