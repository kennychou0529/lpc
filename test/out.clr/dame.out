 solutie : 
1  5  8  6  3  7  2  4  
 solutie : 
1  6  8  3  7  4  2  5  
 solutie : 
1  7  4  6  8  2  5  3  
 solutie : 
1  7  5  8  2  4  6  3  
 solutie : 
2  4  6  8  3  1  7  5  
 solutie : 
2  5  7  1  3  8  6  4  
 solutie : 
2  5  7  4  1  8  6  3  
 solutie : 
2  6  1  7  4  8  3  5  
 solutie : 
2  6  8  3  1  4  7  5  
 solutie : 
2  7  3  6  8  5  1  4  
 solutie : 
2  7  5  8  1  4  6  3  
 solutie : 
2  8  6  1  3  5  7  4  
 solutie : 
3  1  7  5  8  2  4  6  
 solutie : 
3  5  2  8  1  7  4  6  
 solutie : 
3  5  2  8  6  4  7  1  
 solutie : 
3  5  7  1  4  2  8  6  
 solutie : 
3  5  8  4  1  7  2  6  
 solutie : 
3  6  2  5  8  1  7  4  
 solutie : 
3  6  2  7  1  4  8  5  
 solutie : 
3  6  2  7  5  1  8  4  
 solutie : 
3  6  4  1  8  5  7  2  
 solutie : 
3  6  4  2  8  5  7  1  
 solutie : 
3  6  8  1  4  7  5  2  
 solutie : 
3  6  8  1  5  7  2  4  
 solutie : 
3  6  8  2  4  1  7  5  
 solutie : 
3  7  2  8  5  1  4  6  
 solutie : 
3  7  2  8  6  4  1  5  
 solutie : 
3  8  4  7  1  6  2  5  
 solutie : 
4  1  5  8  2  7  3  6  
 solutie : 
4  1  5  8  6  3  7  2  
 solutie : 
4  2  5  8  6  1  3  7  
 solutie : 
4  2  7  3  6  8  1  5  
 solutie : 
4  2  7  3  6  8  5  1  
 solutie : 
4  2  7  5  1  8  6  3  
 solutie : 
4  2  8  5  7  1  3  6  
 solutie : 
4  2  8  6  1  3  5  7  
 solutie : 
4  6  1  5  2  8  3  7  
 solutie : 
4  6  8  2  7  1  3  5  
 solutie : 
4  6  8  3  1  7  5  2  
 solutie : 
4  7  1  8  5  2  6  3  
 solutie : 
4  7  3  8  2  5  1  6  
 solutie : 
4  7  5  2  6  1  3  8  
 solutie : 
4  7  5  3  1  6  8  2  
 solutie : 
4  8  1  3  6  2  7  5  
 solutie : 
4  8  1  5  7  2  6  3  
 solutie : 
4  8  5  3  1  7  2  6  
 solutie : 
5  1  4  6  8  2  7  3  
 solutie : 
5  1  8  4  2  7  3  6  
 solutie : 
5  1  8  6  3  7  2  4  
 solutie : 
5  2  4  6  8  3  1  7  
 solutie : 
5  2  4  7  3  8  6  1  
 solutie : 
5  2  6  1  7  4  8  3  
 solutie : 
5  2  8  1  4  7  3  6  
 solutie : 
5  3  1  6  8  2  4  7  
 solutie : 
5  3  1  7  2  8  6  4  
 solutie : 
5  3  8  4  7  1  6  2  
 solutie : 
5  7  1  3  8  6  4  2  
 solutie : 
5  7  1  4  2  8  6  3  
 solutie : 
5  7  2  4  8  1  3  6  
 solutie : 
5  7  2  6  3  1  4  8  
 solutie : 
5  7  2  6  3  1  8  4  
 solutie : 
5  7  4  1  3  8  6  2  
 solutie : 
5  8  4  1  3  6  2  7  
 solutie : 
5  8  4  1  7  2  6  3  
 solutie : 
6  1  5  2  8  3  7  4  
 solutie : 
6  2  7  1  3  5  8  4  
 solutie : 
6  2  7  1  4  8  5  3  
 solutie : 
6  3  1  7  5  8  2  4  
 solutie : 
6  3  1  8  4  2  7  5  
 solutie : 
6  3  1  8  5  2  4  7  
 solutie : 
6  3  5  7  1  4  2  8  
 solutie : 
6  3  5  8  1  4  2  7  
 solutie : 
6  3  7  2  4  8  1  5  
 solutie : 
6  3  7  2  8  5  1  4  
 solutie : 
6  3  7  4  1  8  2  5  
 solutie : 
6  4  1  5  8  2  7  3  
 solutie : 
6  4  2  8  5  7  1  3  
 solutie : 
6  4  7  1  3  5  2  8  
 solutie : 
6  4  7  1  8  2  5  3  
 solutie : 
6  8  2  4  1  7  5  3  
 solutie : 
7  1  3  8  6  4  2  5  
 solutie : 
7  2  4  1  8  5  3  6  
 solutie : 
7  2  6  3  1  4  8  5  
 solutie : 
7  3  1  6  8  5  2  4  
 solutie : 
7  3  8  2  5  1  6  4  
 solutie : 
7  4  2  5  8  1  3  6  
 solutie : 
7  4  2  8  6  1  3  5  
 solutie : 
7  5  3  1  6  8  2  4  
 solutie : 
8  2  4  1  7  5  3  6  
 solutie : 
8  2  5  3  1  7  4  6  
 solutie : 
8  3  1  6  2  5  7  4  
 solutie : 
8  4  1  3  6  2  7  5  
